### PR TITLE
Fixes type-hinting by using contract/interface class instead of actual implementation class

### LINKED ISSUE #27825: Argument typehint should use contract class instead of implementaiton class
- Laravel Version: >=5.4.*
- PHP Version: 7.*
- Database Driver & Version: Not applicable

### Description:
I am working on a package where I extend 'queue' service following way, where QueueManager is some class of  mine, lets say which of course implements the contarct -

```php
$this->app->extend(
    'queue',
    function (\Illuminate\Contracts\Queue\Factory $factory) use ($config) {
        return new QueueManager($factory);
    }
);
```

Doing this causes an issue where laravel's Queue/Worker's constructor has arg type-hinting for actual implementation instead of the contract. Below is that line for reference -

https://github.com/laravel/framework/blob/0bb3d37033fb45a4f1a1dba48cf55ed15b24ab55/src/Illuminate/Queue/Worker.php#L60-L70

### Steps To Reproduce:

Explained as part of description ^ above.

### Resolution:

Suggesting to use `\Illuminate\Contracts\Queue\Factory` instead of `\Illuminate\Queue\QueueManager` for the argument type hinting at above place i.e. Worker.php

I will file a pr as it is not much effort, but will wait for comments.

### PR DESCRIPTION
**Pull request by jitendra-1217**
Fixes https://github.com/laravel/framework/issues/27825

It will help extending 'queue' service of Laravel's framework possible. There is details of issue etc on linked issue above.

### CODE REVIEW

**Comment by jmarcher** [comment]
This is incorrect, `Illuminate\Queue\QueueManager` also implements `Illuminate\Contracts\Queue\Monitor`
I think we are better off this way type-hinting `Illuminate\Queue\QueueManager`

**Comment by X-Coder264** [comment]
I agree with @jitendra-1217. If there is a contract the classes in the framework should typehint that contract instead of the concrete implementation as that causes issues like #27825 and defeat the entire point of having the contract in the first place.

The `Monitor` interface and its methods aren't needed for the worker class as far as I can see.

File: src/Illuminate/Queue/Worker.php [hunk 1]
```diff
@@ -617,7 +618,7 @@ public function getManager()
     /**
      * Set the queue manager instance.
      *
-     * @param  \Illuminate\Queue\QueueManager  $manager
+     * @param  \Illuminate\Contracts\Queue\Factory  $manager
      * @return void
      */
     public function setManager(QueueManager $manager)
```
Comment (X-Coder264) [inline src/Illuminate/Queue/Worker.php]: You changed the docblock, but you forgot to change the actual method typehint.

**Comment by jmarcher** [comment]
Even if the `Monitor` interface is not needed for the worker, this a breaking change.

**Comment by X-Coder264** [comment]
I agree, this should definitively target `master`.

File: src/Illuminate/Queue/Worker.php (hunk 1 above)
Comment (jitendra-1217) [reply src/Illuminate/Queue/Worker.php]: Thanks for pointing.

**Comment by driesvints** [comment]
This breaks the method signature on a point release.

**Comment by jitendra-1217** [comment]
Thanks @jmarcher , @X-Coder264 for your comments. Understood or say reiterated/realized a lot of things here.

**Comment by jitendra-1217** [comment]
> The change (changing the typehint from a concrete implementation to a contract) is good, but further work is needed to refactor this so that only the methods that exist on that contract get called in the Worker class.

I would want to do this. Because it is good to have way to extend queue service. There are lot of public methods in QueueManager class and contract-ifying will take more thought.

**Comment by X-Coder264** [comment]
@jitendra-1217 Sure. It would be great to get such a PR targeting `master` so that it gets released in 5.9.

**Comment by X-Coder264** [comment]
@jitendra-1217 What do you mean by 

> There are lot of public methods in QueueManager class and contract-ifying will take more thought.

The only method from the `QueueManager` that is called in the `Worker` class is `isDownForMaintenance`. The easiest way to solve that is by injecting the `Application` contract into the `Worker` too.

**Pull request closed without merging**